### PR TITLE
Update srpc_generator

### PR DESCRIPTION
**Pull request by holmes1412**
Add two updates mentioned in this Pull Request: https://github.com/holmes1412/srpc/pull/26

(1) Some redundancy in specifying the input IDL file type
(2) Crash if output directory doesn't exist

Also make it compatible with the original usage and add a version check.

Correct usage: 
```sh
srpc_generator protobuf echo_pb.proto ./
```

```sh
srpc_generator echo_thrift.thrift ./
```

Error usage and message:
```sh
liyingxin@P_QZLIU-MB0 srpc_1412_new % srpc_generator .xxx ./                             
ERROR: Invalid IDL file .xxx
Usage:
	srpc_generator <idl_file> <output_dir>
```

```sh
liyingxin@P_QZLIU-MB0 srpc_1412_new % srpc_generator xxx tutorial/echo_pb.proto ./
ERROR: Invalid IDL type xxx
Usage:
	srpc_generator <idl_file> <output_dir>
```

```sh
liyingxin@P_QZLIU-MB0 srpc_1412_new % srpc_generator --version                    
srpc_generator version 0.9.5
```

```sh
liyingxin@P_QZLIU-MB0 srpc_1412_new % srpc_generator          
Usage:
	srpc_generator <idl_file> <output_dir>
```

### CODE REVIEW

**Comment by holmes1412** [comment]
@linqigang888 I may need your help to review this Pull Request.
I also wanted to choose you as a reviewer at the REVIEWERS column on the right, but I cannot find your account on the column list.

**Comment by holmes1412** [comment]
According to Lance's suggestion,  optargs and getopt_long can make order of arguments won't matter.
I will implement this feature next time.

File: src/generator/compiler.cc [hunk 1]
```diff
@@ -21,24 +21,69 @@
...
+		fprintf(stderr, SRPC_GENERATOR_USAGE, argv[0]);
 		return 0;
 	}
 
-	char *idl = const_cast<char *>(argv[1]);
-	bool is_thrift = (strcasecmp(idl, "thrift") == 0);
-	const char *idl_file = argv[2];
+	if (argc == 4)
+	{
+		if (strcmp(argv[1], "protobuf") && strcmp(argv[1], "thrift"))
```
Comment (Barenboim) [inline src/generator/compiler.cc:64]: 改成strcasecmp吧

File: src/generator/compiler.cc [hunk 2]
```diff
@@ -21,24 +21,69 @@
...
+{
+	TYPE_UNKNOWN = -1,
+	TYPE_PROTOBUF = 0,
+	TYPE_THRIFT = 1,
+};
+
+const char *SRPC_VERSION = "0.9.5";
+
+static int check_idl_type(const char *filename);
+const char *SRPC_GENERATOR_USAGE = "Usage:\n\t%s <idl_file> <output_dir>\n";
```
Comment (Barenboim) [inline src/generator/compiler.cc:38]: 这个USAGE改一下，中间加上[protobuf/thrift]这个可选参数

File: src/generator/compiler.cc [hunk 3]
```diff
@@ -21,24 +21,69 @@
...
+		{
+			fprintf(stderr, "ERROR: Invalid IDL type %s\n", argv[1]);
+			fprintf(stderr, SRPC_GENERATOR_USAGE, argv[0]);
+			return 0;
+		}
+
+		idl_file_id++;
+	}
+
+	idl_type = check_idl_type(argv[idl_file_id]);
```
Comment (Barenboim) [inline src/generator/compiler.cc:74]: 如果用户填写了protobuf或thrift，应该就不再判断文件名吧？

File: src/generator/compiler.cc (hunk 1 above)
Comment (holmes1412) [reply src/generator/compiler.cc]: ok

File: src/generator/compiler.cc (hunk 2 above)
Comment (holmes1412) [reply src/generator/compiler.cc]: 改好了。现在的usage是：
```
srpc_generator [protobuf|thrift] <idl_file> <output_dir>
```

File: src/generator/compiler.cc (hunk 3 above)
Comment (holmes1412) [reply src/generator/compiler.cc]: 对的，改好了。

**Comment by linqigang888** [comment]
Hello,

I think these changes are good! Here are a few comments:

(1) I like the use of the enum for file type, it is more "complete".

(2) I understand using the file extension for determining the IDL type. This can work. Do we assume users will always have the correct file type? That is to say, someone would not realise that their file was .proto but it was named .thrift instead? I think this assumption is fine as your audience is developers and they should know better. You should make sure that srpc_generator exits gracefully in the case of the wrong input file type (extension mismatch) for this situation.

In general, I assume the worst of users and write software to protect against them :)

(3) Future suggestion: my code added a guard against a crash when the output directory did not exist. Perhaps we could create the output directory for the user, should it not exist? I was able to find a cross-platform solution to test if a directory exists (Linux/Windows) but didn't have the time to run down the same code to create a directory in a cross-platform fashion.

(4) When I made my initial PR, it was a "suggestion" type so I'm glad to see you implemented a way to use both the argc "4" version as well as the newer "3" version. This is great for stability and backwards compatibility. If you do implement getopt_long or optargs, I would then revise the code to accept the "4" version (no optarg/getopt_long) or the optarg/getopt_long version for backwards compatibility/legacy support. Once everything is updated, I would fully switch to getopt.

Best,

Lance

**Comment by holmes1412** [comment]
@linqigang888 Thank you for your comment!  

(2) About the file extension - Yes, people will realize what their file is. Because the usage of the file extension is the original convention of protobuf and thrift, which means protobuf users can make sure their idl file end with .proto and so as the thrift users.
When the idl type does not match the file extension, srpc_generator will not crash, but exist with error message instead.

(3)  create non-existent directory - This is a good suggestion, but I have the `protoc` and it has not support this feature yet. So I think I will discuss with my colleagues and maybe implement this in the future.

(4) I think argc "3" version is better and less redundancy ~ I will finish the optarg/getopt_long version within this week, the usage will be like: `srpc_generator -i <IDL_FILE> -o <OUTPUT_DIR>`. But it will also support the old argc "4" version: `srpc_generator [protobuf|thirft] <IDL_FILE> <OUTPUT_DIR>`. Is this ok? @Barenboim

**Review by Barenboim — CHANGES_REQUESTED** [review]
不要引入stat，windows下没有。其实无需判断目录是否存在，直接创建文件查看返回值就可以。
判断文件扩展名的代码应该用strcmp而不是strstr，那个判断逻辑是错的。

**Review by Barenboim — APPROVED** [review]

**Pull request merged**